use cargo dist for binary releases

webern opened the pull request:
*Issue #, if available:*

Closes #48

*Description of changes:*

Produce a new GitHub Actions workflow for publishing releases based on tags and adding binary builds to the releases.

*Testing*

This was tested extensively on my fork and it worked:
- https://github.com/webern/twoliter/releases/tag/v0.0.3-20230824115213

There will be a fair amount of tag churn getting this working for the first time. These will be `v0.0.3-rc1` etc. Once everything is working a `v0.0.3` tag will be produced and the rc's can be deleted from the git repo to clean things up. I will write up a release procedure in a maintainers section of the README but will need some experience with it first.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

webern committed:
release v0.0.3

bcressey commented on CHANGELOG.md:
```diff
@@ -21,6 +38,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
-[unreleased]: https://github.com/bottlerocket-os/twoliter/compare/v0.0.2...HEAD
-[0.0.2]: https://github.com/bottlerocket-os/twoliter/compare/v0.0.1...v0.0.2
-[0.0.1]: https://github.com/bottlerocket-os/twoliter/releases/tag/v0.0.1
+[unreleased]: https://github.com/webern/twoliter/compare/v0.0.3...HEAD
+[0.0.3]: https://github.com/webern/twoliter/compare/v0.0.2...v0.0.3
+[0.0.2]: https://github.com/webern/twoliter/compare/v0.0.1...v0.0.2
+[0.0.1]: https://github.com/webern/twoliter/releases/tag/v0.0.1
```
Should this be `bottlerocket-os/twoliter` or is this going to change in the forthcoming tag churn?

webern replied:
Good catch. This is a mistake.

bcressey commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,163 @@
+        run: ${{ matrix.install-dist }}
+      - name: Install Cross
+        run: |
+          cargo install cross --git https://github.com/cross-rs/cross
```
Should we pin to a specific version here?

bcressey commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,163 @@
+        run: |
+          export PATH=$CARGO_HOME/bin:$PATH
+          # Actually do builds and make zips and whatnot
+          cargo dist build --verbose=debug --tag=${{ github.ref_name }} --output-format=json ${{ matrix.dist-args }} > dist-manifest.json
```
I think it would help future rebases to split this into two commits:
* one for the unmodified `cargo dist` workflow
* one for the local changes we've made to that workflow

That way, the local changes can be picked out and potentially reapplied if we want to rebase to a newer cargo dist release for new features or something.

webern committed:
initialize cargo dist

Run cargo dist init on the repository. This gives us new metadata in the
workspace Cargo.toml and a new GitHub Actions file, release.yml.

webern committed:
customize cargo dist

Change cargo dist settings to give us the Rust targets that we want
(musl for both x86_64 and aarch64). Change release.yml to support cross
compilation.

bcressey reviewed (approved):
nit: "customize cargo dist" commit doesn't mention aarch64
